File: backend/processor/src/main/java/com/middleware/processor/service/interfaces/ProcessedFileService.java
package com.middleware.processor.service.interfaces;
    
import com.middleware.shared.model.ProcessedFile;
import com.middleware.shared.model.Interface;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.PageRequest;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
    
/**
 * Service interface for processed file operations.
 */
public interface ProcessedFileService {
    /**
     * Get all processed files with pagination.
     *
     * @param pageable The pagination information
     * @return Page of processed files
     */
    Page<ProcessedFile> getAllProcessedFiles(Pageable pageable);

    /**
     * Get processed files by client with pagination.
     *
     * @param clientId The ID of the client
     * @param pageable The pagination information
     * @return Page of processed files for the client
     */
    Page<ProcessedFile> getProcessedFilesByClient(Long clientId, Pageable pageable);

    /**
     * Get processed files by interface with pagination.
     *
     * @param interfaceId The ID of the interface
     * @param pageable The pagination information
     * @return Page of processed files for the interface
     */
    Page<ProcessedFile> getProcessedFilesByInterface(Long interfaceId, Pageable pageable);

    /**
     * Get processed files by date range with pagination.
     *
     * @param startDate The start date
     * @param endDate The end date
     * @param pageable The pagination information
     * @return Page of processed files within the date range
     */
    Page<ProcessedFile> getProcessedFilesByDateRange(LocalDateTime startDate, LocalDateTime endDate, Pageable pageable);

    /**
     * Get processed files by status with pagination.
     *
     * @param status The status to filter by
     * @param pageable The pagination information
     * @return Page of processed files with the specified status
     */
    Page<ProcessedFile> getProcessedFilesByStatus(String status, Pageable pageable);

    /**
     * Get a processed file by ID.
     *
     * @param id The ID of the processed file
     * @return Optional containing the processed file if found
     */
    Optional<ProcessedFile> getProcessedFileById(Long id);

    /**
     * Create a new processed file.
     *
     * @param processedFile The processed file to create
     * @return The created processed file
     */
    ProcessedFile createProcessedFile(ProcessedFile processedFile);

    /**
     * Update an existing processed file.
     *
     * @param id The ID of the processed file to update
     * @param processedFile The updated processed file data
     * @return The updated processed file
     */
    ProcessedFile updateProcessedFile(Long id, ProcessedFile processedFile);

    /**
     * Delete a processed file.
     *
     * @param id The ID of the processed file to delete
     */
    void deleteProcessedFile(Long id);

    /**
     * Get all error files with pagination.
     *
     * @param pageable The pagination information
     * @return Page of error files
     */
    Page<ProcessedFile> getErrorFiles(Pageable pageable);

    /**
     * Get error files by date range with pagination.
     *
     * @param startDate The start date
     * @param endDate The end date
     * @param pageable The pagination information
     * @return Page of error files within the date range
     */
    Page<ProcessedFile> getErrorFilesByDateRange(LocalDateTime startDate, LocalDateTime endDate, Pageable pageable);

    /**
     * Get processed files with advanced filtering and pagination.
     *
     * @param page The page number
     * @param size The page size
     * @param sortBy The field to sort by
     * @param direction The sort direction
     * @param fileNameFilter The search term
     * @param statusFilter The status to filter by
     * @param startDate The start date
     * @param endDate The end date
     * @param clientId The client ID to filter by
     * @param interfaceId The interface ID to filter by
     * @return Page of processed files matching the criteria
     */
    Page<ProcessedFile> getProcessedFiles(int page, int size, String sortBy, String direction,
                                        String fileNameFilter, String statusFilter,
                                        LocalDateTime startDate, LocalDateTime endDate,
                                        Long clientId, Long interfaceId);

    /**
     * Get processed files by client with pagination and sorting.
     *
     * @param clientId The ID of the client
     * @param page The page number
     * @param size The page size
     * @param sortBy The field to sort by
     * @param sortDirection The sort direction
     * @return Page of processed files for the client
     */
    Page<ProcessedFile> getProcessedFilesByClient(Long clientId, int page, int size, String sortBy, String sortDirection);

    /**
     * Get processed files by client and status with pagination and sorting.
     *
     * @param clientId The ID of the client
     * @param status The status to filter by
     * @param page The page number
     * @param size The page size
     * @param sortBy The field to sort by
     * @param sortDirection The sort direction
     * @return Page of processed files for the client with the specified status
     */
    Page<ProcessedFile> getProcessedFilesByClientAndStatus(Long clientId, String status, 
                                                         int page, int size, String sortBy, String sortDirection);

    /**
     * Get processed files by client and date range with pagination and sorting.
     *
     * @param clientId The ID of the client
     * @param startDate The start date
     * @param endDate The end date
     * @param page The page number
     * @param size The page size
     * @param sortBy The field to sort by
     * @param sortDirection The sort direction
     * @return Page of processed files for the client within the date range
     */
    Page<ProcessedFile> getProcessedFilesByClientAndDateRange(Long clientId, LocalDateTime startDate, LocalDateTime endDate,
                                                            int page, int size, String sortBy, String sortDirection);

    /**
     * Get processed files by status.
     *
     * @param status The status to filter by
     * @return List of processed files with the specified status
     */
    List<ProcessedFile> getProcessedFilesByStatus(String status);

    Page<ProcessedFile> getProcessedFilesByClient(Long clientId, PageRequest pageRequest);

    Page<ProcessedFile> searchProcessedFiles(String fileName, PageRequest pageRequest);

    Page<ProcessedFile> getProcessedFilesByStatus(String status, PageRequest pageRequest);

    Page<ProcessedFile> getProcessedFilesByDateRange(LocalDateTime startDate, LocalDateTime endDate, PageRequest pageRequest);

    Page<ProcessedFile> getProcessedFilesByClientAndStatus(Long clientId, String status, PageRequest pageRequest);

    Page<ProcessedFile> getProcessedFilesByClientAndDateRange(Long clientId, LocalDateTime startDate, LocalDateTime endDate, PageRequest pageRequest);

    /**
     * Find a processed file by filename and interface ID.
     *
     * @param fileName The name of the file
     * @param interfaceId The ID of the interface
     * @return Optional containing the processed file if found
     */
    Optional<ProcessedFile> findByFileNameAndInterfaceId(String fileName, Long interfaceId);

    /**
     * Find the most recent processed file by filename and interface ID.
     *
     * @param fileName The name of the file
     * @param interfaceId The ID of the interface
     * @return Optional containing the most recent processed file if found
     */
    Optional<ProcessedFile> findMostRecentByFileNameAndInterfaceId(String fileName, Long interfaceId);

    /**
     * Find processed files with various filters and pagination.
     *
     * @param clientId The client ID to filter by
     * @param interfaceId The interface ID to filter by
     * @param status The status to filter by
     * @param fileName The filename to search for
     * @param startDate The start date for date range filtering
     * @param endDate The end date for date range filtering
     * @param pageable The pagination information
     * @return Page of processed files matching the criteria
     */
    Page<ProcessedFile> findProcessedFiles(
            Long clientId, 
            Long interfaceId, 
            String status, 
            String fileName, 
            LocalDateTime startDate, 
            LocalDateTime endDate,
            Pageable pageable);

    /**
     * Find an existing processed file or create a new one if it doesn't exist.
     * This method ensures atomic find-or-create operation with proper transaction handling.
     *
     * @param fileName The name of the file
     * @param interfaceEntity The interface entity
     * @param defaultFile The default file to create if none exists
     * @return The existing or newly created processed file
     */
    ProcessedFile findOrCreateProcessedFile(String fileName, Interface interfaceEntity, ProcessedFile defaultFile);

    /**
     * Atomically update a processed file's status and content.
     * This method ensures that updates are performed on the most recent version of the record.
     *
     * @param id The ID of the processed file
     * @param status The new status
     * @param errorMessage The error message (if any)
     * @param content The file content (if any)
     * @return The updated processed file
     */
    ProcessedFile atomicUpdateProcessedFile(Long id, String status, String errorMessage, String content);
}

File: frontend/src/services/inbound_config/processedFileService.ts
import { api } from '../core/apiService';
import { ProcessedFile } from './types';
import { PageResponse } from '../core/types';
import { handleApiError } from '../../utils/errorHandler';

export const processedFileService = {
  uploadFile: async (file: File, interfaceId: number): Promise<ProcessedFile> => {
    try {
      const formData = new FormData();
      formData.append('file', file);

      const response = await api.post<ProcessedFile>(
        `/api/files/upload/${interfaceId}`,
        formData,
        {
          headers: {
            'Content-Type': 'multipart/form-data'
          },
          timeout: 60000 // 60 seconds timeout for large files
        }
      );
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getAllProcessedFiles: async (
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc',
    fileNameFilter?: string,
    statusFilter?: string,
    startDate?: string,
    endDate?: string,
    clientId?: number,
    interfaceId?: number
  ): Promise<PageResponse<ProcessedFile>> => {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        size: size.toString(),
        sortBy,
        direction
      });
      
      if (fileNameFilter) params.append('fileNameFilter', fileNameFilter);
      if (statusFilter) params.append('statusFilter', statusFilter);
      if (startDate) params.append('startDate', startDate);
      if (endDate) params.append('endDate', endDate);
      if (clientId) params.append('clientId', clientId.toString());
      if (interfaceId) params.append('interfaceId', interfaceId.toString());
      
      const response = await api.get<PageResponse<ProcessedFile>>('/api/processed-files', { params });
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getProcessedFileById: async (id: number): Promise<ProcessedFile> => {
    try {
      const response = await api.get<ProcessedFile>(`/api/files/${id}`);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  createProcessedFile: async (processedFileData: Omit<ProcessedFile, 'id'>): Promise<ProcessedFile> => {
    const response = await api.post<ProcessedFile>(`/api/processed-files`, processedFileData);
    return response.data;
  },

  updateProcessedFile: async (id: number, processedFileData: Partial<ProcessedFile>): Promise<ProcessedFile> => {
    const response = await api.put<ProcessedFile>(`/api/processed-files/${id}`, processedFileData);
    return response.data;
  },

  deleteProcessedFile: async (id: number): Promise<void> => {
    try {
      await api.delete(`/api/files/${id}`);
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getProcessedFilesByClient: async (
    clientId: number,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await api.get<PageResponse<ProcessedFile>>(`/api/processed-files/client/${clientId}?${params.toString()}`);
    return response.data;
  },

  searchProcessedFiles: async (
    fileName: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('fileName', fileName);
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await api.get<PageResponse<ProcessedFile>>(`/api/processed-files/search?${params.toString()}`);
    return response.data;
  },

  getProcessedFilesByStatus: async (
    status: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await api.get<PageResponse<ProcessedFile>>(`/api/processed-files/status/${status}?${params.toString()}`);
    return response.data;
  },

  getProcessedFilesByDateRange: async (
    startDate: string,
    endDate: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('startDate', startDate);
    params.append('endDate', endDate);
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await api.get<PageResponse<ProcessedFile>>(`/api/processed-files/date-range?${params.toString()}`);
    return response.data;
  },

  getProcessedFilesByClientAndStatus: async (
    clientId: number,
    status: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await api.get<PageResponse<ProcessedFile>>(`/api/processed-files/client/${clientId}/status/${status}?${params.toString()}`);
    return response.data;
  },

  getProcessedFilesByClientAndDateRange: async (
    clientId: number,
    startDate: string,
    endDate: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'processedDate',
    direction: string = 'desc'
  ): Promise<PageResponse<ProcessedFile>> => {
    let params = new URLSearchParams();
    params.append('startDate', startDate);
    params.append('endDate', endDate);
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('direction', direction);
    
    const response = await api.get<PageResponse<ProcessedFile>>(`/api/processed-files/client/${clientId}/date-range?${params.toString()}`);
    return response.data;
  },

  downloadFile: async (id: number): Promise<Blob> => {
    try {
      const response = await api.get(`/api/files/${id}/download`, {
        responseType: 'blob'
      });
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  reprocessFile: async (id: number): Promise<void> => {
    try {
      await api.post(`/api/files/reprocess/${id}`);
    } catch (error) {
      throw handleApiError(error);
    }
  }
}; 